File: dsl-api/src/main/java/cbs/dsl/api/TransactionDefinition.java
package cbs.dsl.api;

import cbs.dsl.api.TransactionTypes.TransactionInput;
import cbs.dsl.api.TransactionTypes.TransactionOutput;

/**
 * Defines a transaction — a unit of work with preview, execute, and rollback phases.
 *
 * <p>Implementations are typically created via the Kotlin DSL {@code transaction { }} block or
 * annotated with {@link DslComponent} for compile-time registration.
 */
@FunctionalInterface
public interface TransactionDefinition extends StandardDslDefinition {

  TransactionOutput execute(TransactionInput input);

  default TransactionOutput rollback(TransactionInput input) {
    return input.asOutput();
  }
}

File: dsl-api/src/main/kotlin/cbs/dsl/api/TransactionDefinition.kt
package cbs.dsl.api

import cbs.dsl.api.context.TransactionContext

interface TransactionDefinition {
  val code: String

  /**
   * Optional display name for this transaction. Used to distinguish DSL overrides from the
   * underlying implementation class/bean identified by [code]. When set, the DSL block is treated
   * as a named override of the bean registered under [code].
   *
   * Example:
   * ```kotlin
   * transaction("KYC_CHECK") {
   *     name("TestKycCheck")
   *     execute { ctx -> ctx["kycVerified"] = true }
   * }
   * ```
   */
  val name: String?
    get() = null

  /**
   * List of parameter definitions declared in the `parameters { }` block. Used for validation and
   * documentation purposes.
   */
  val parameters: List<ParameterDefinition>
    get() = emptyList()

  /**
   * Optional context enrichment block that runs before each phase block (preview, execute,
   * rollback). Allows transactions to enrich the context with additional data before execution.
   *
   * Example:
   * ```kotlin
   * transaction("KYC_CHECK") {
   *     context { ctx ->
   *         ctx["kycProvider"] = "internal"
   *     }
   *     execute { ctx -> ctx["kycVerified"] = true }
   * }
   * ```
   */
  val contextBlock: (TransactionContext) -> Unit
    get() = {}

  fun preview(input: TransactionInput): TransactionOutput

  fun execute(input: TransactionInput): TransactionOutput

  fun rollback(input: TransactionInput): TransactionOutput
}
